File: 200. Backtrack - Number of Islands.py
"""
  I: matrix (array of arrays representing grid)
  O: number of islands connecting
  C:
  E:

  Strategy:
    - Iterate through matrix
    - If value = 1, incremement, then recurse through all edges, setting each equal to zero
    - continue iterating until end of matrix
"""


class Solution:
    def numIslands(self, grid) -> int:
        count = 0

        def checkEdges(r, c):
            # Base:
            # Return if:
            # 1. square = 0
            # 2. i or j are outside the bounds (and therefore are water)
            if (
                (r < 0 or r >= len(grid))
                or (c < 0 or c >= len(grid[r]))
                or (grid[r][c] == "0")
            ):
                return

            # Set grid to zero
            grid[r][c] = "0"

            # Recursive on each edge
            checkEdges(r - 1, c)  # up
            checkEdges(r + 1, c)  # down
            checkEdges(r, c + 1)  # right
            checkEdges(r, c - 1)  # left

        # For each row
        for row in range(0, len(grid)):
            # For each column
            for col in range(0, len(grid[row])):
                # If 0, then continue
                if grid[row][col] == "0":
                    continue

                # If 1, then increment and check all edges
                count += 1
                checkEdges(row, col)

        return count


# ==== TESTING ====
# Input: grid = [
#   ["1","1","1","1","0"],
#   ["1","1","0","1","0"],
#   ["1","1","0","0","0"],
#   ["0","0","0","0","0"]
# ]
# Output: 1
# print(Solution.numIslands(None, [
#   ["1","1","1","1","0"],
#   ["1","1","0","1","0"],
#   ["1","1","0","0","0"],
#   ["0","0","0","0","0"]
# ]))

print(
    Solution.numIslands(
        None,
        [
            ["1", "1", "0", "0", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "1", "0", "0"],
            ["0", "0", "0", "1", "1"],
        ],
    )
)

File: 200. Backtrack - Number of Islands.js
/*
  Problem:

  Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

  An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

  I: m x n matrix of 1s and 0s
  O: int - number of islands
  C:
  E:

  Strategy:
    - Loop through each tile
    - When we find an island, increment counter, and recurse on edges
    - Set each island tile to zero when we're finished
*/

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  let count = 0;
  let checkIslandEdges = (r, c) => {
    // Base
    if (r < 0 || r >= grid.length) {
      return;
    }
    if (c < 0 || c >= grid[r].length) {
      return;
    }
    if (grid[r][c] === '0') {
      return;
    }

    // Recursive
    grid[r][c] = '0';
    checkIslandEdges(r + 1, c);
    checkIslandEdges(r - 1, c);
    checkIslandEdges(r, c + 1);
    checkIslandEdges(r, c - 1);
  };

  for (var row = 0; row < grid.length; row++) {
    for (var col = 0; col < grid[row].length; col++) {
      if (grid[row][col] === '0') {
        continue;
      }
      count++;
      checkIslandEdges(row, col);
    }
  }

  return count;
};

// ===== TESTING =====
// Input: grid = [
//   ["1","1","1","1","0"],
//   ["1","1","0","1","0"],
//   ["1","1","0","0","0"],
//   ["0","0","0","0","0"]
// ]
// Output: 1
console.log(
  numIslands([
    ['1', '1', '1', '1', '0'],
    ['1', '1', '0', '1', '0'],
    ['1', '1', '0', '0', '0'],
    ['0', '0', '0', '0', '0'],
  ])
);

// Example 2:

// Input: grid = [
//   ["1","1","0","0","0"],
//   ["1","1","0","0","0"],
//   ["0","0","1","0","0"],
//   ["0","0","0","1","1"]
// ]
// Output: 3
console.log(
  numIslands([
    ['1', '1', '0', '0', '0'],
    ['1', '1', '0', '0', '0'],
    ['0', '0', '1', '0', '0'],
    ['0', '0', '0', '1', '1'],
  ])
);
